app/src/main/java/fr/free/nrw/commons/upload/PendingUploadsFragment.kt

```diff
@@ -0,0 +1,297 @@
+                binding.pendingUplaodsLl.visibility = View.GONE
+                uploadProgressActivity.hidePendingIcons()
+            } else {
+                if (totalUploads == 0){
+                    totalUploads = contributionsSize
+                    binding.progressBarPending.max = totalUploads
+                }
+                binding.nopendingTextView.visibility = View.GONE
+                binding.pendingUplaodsLl.visibility = View.VISIBLE
+
+                val sortedContributionsList: List<Contribution> = if (VERSION.SDK_INT >= VERSION_CODES.N) {
+                    contributionsList.sortedByDescending { it.dateModifiedInMillis() }
+                } else {
+                    contributionsList.sortedBy { it.dateModifiedInMillis() }.reversed()
+                }
+
+                val newContributionList: MutableList<Contribution> = sortedContributionsList.toMutableList()
+                val listOfRemoved: MutableList<Contribution> = mutableListOf()
+                val last = sortedContributionsList.last()
+                for (i in sortedContributionsList.indices) {
+                    val current = sortedContributionsList[i]
+                    if (current.transferred == 0L && (current.dateModifiedInMillis() / 100) > (last.dateModifiedInMillis() / 100)){
+                        listOfRemoved.add(current)
+                    }
+                }
+                newContributionList.removeAll(listOfRemoved)
+                newContributionList.addAll(listOfRemoved)
+                val adapter = PendingUploadsAdapter(newContributionList, this)
+                binding.pendingUploadsRecyclerView.setAdapter(adapter)
+                binding.progressTextView.setText((totalUploads-contributionsSize).toString() + "/" + totalUploads + " uploaded")
+                binding.progressBarPending.progress = totalUploads-contributionsSize
+                if (pausedOrQueuedUploads == contributionsSize) {
+                    uploadProgressActivity.setPausedIcon(true)
+                }else{
+                    uploadProgressActivity.setPausedIcon(false)
+                }
+            }
+        }
+    }
+
+    override fun deleteUpload(contribution: Contribution?) {
+        showAlertDialog(
+            requireActivity(),
+            String.format(
+                Locale.getDefault(),
+                getString(R.string.cancelling_upload)
+            ),
+            String.format(
+                Locale.getDefault(),
+                getString(R.string.cancel_upload_dialog)
+            ),
+            String.format(Locale.getDefault(), getString(R.string.yes)),
+            String.format(Locale.getDefault(), getString(R.string.no)),
+            {
+                ViewUtil.showShortToast(context, R.string.cancelling_upload)
+                pendingUploadsPresenter.deleteUpload(contribution, this.requireContext().applicationContext)
+                CommonsApplication.cancelledUploads.add(contribution!!.pageId)
```

Comment:
Are we keeping the hashset for some cases?

Reply:
Ah sorry! I forgot to remove here, will share a newer commit to solve it

Reply:
@kanahia1 One more question: are we handling UI changes in case of cancelled uploads using LiveData? If not, do we really require an observable query?

Reply:
We are handling UI using livedata